WorkspaceServer/Packaging/Package.cs

```diff
@@ -575,7 +575,25 @@ protected virtual bool ShouldDoFullBuild()
         protected virtual bool ShouldDoDesignTimeBuild()
         {
             return DesignTimeBuildResult == null
-                   || DesignTimeBuildResult.Succeeded == false;
+                   || DesignTimeBuildResult.Succeeded == false
+                   || FilesHaveBeenAddedOrDeleted();
+        }
+
+        protected virtual bool FilesHaveBeenAddedOrDeleted()
+        {
+            if (DesignTimeBuildResult != null)
+            {
+                var filesInDirectory = Directory.GetFiles("*.cs").Select(file => file.FullName);
+                var documentsInLastBuild = DesignTimeBuildResult.GetWorkspace().CurrentSolution.Projects.First().Documents.Select(document => document.FilePath);
+                if (filesInDirectory.Count() != documentsInLastBuild.Count())
+                {
+                    return true;
+                }
+
+                return !documentsInLastBuild.SequenceEqual(filesInDirectory);
```

Comment:
I don't think you can assume that the documents are represented in the same order in the workspace and in the file system.